convert-gptneox-hf-to-gguf.py

```diff
@@ -111,7 +111,7 @@ def count_model_parts(dir_model: str) -> int:
 
 print("gguf: get tokenizer metadata")
 
-tokens: List[str] = []
+tokens: List[bytearray] = []
```

Comment:
These changes shut up the type warnings but I'm not 100% sure they're correct. The alternative would be to leave it `List[str]` and then convert the token text to a string. I _assume_ it's already UTF-8 bytes so there probably isn't a functional difference.

Reply:
The types are correct this way, because when we get to `tokens.append(text)`, 'text' is explicitly a bytearray.

Reply:
> The types are correct this way, because when we get to tokens.append(text), 'text' is explicitly a bytearray.

Right. I meant my change fixes the type warning but the part I wasn't sure about was whether `text` _actually_ is supposed to be a `bytearray` there and what is supposed to get submitted to `gguf` to write out the tokens. The type annotation for `add_token_list` method is also just `List` and doesn't specify what the element is supposed to be.

Reply:
The decision was  made to accept several types as input to the token list depending on type of he tokenizer output (spm vs bpe)
https://github.com/ggerganov/llama.cpp/blob/dd0dc366dab10e8df28d3924e7f313b5c695e908/gguf-py/gguf/gguf.py#L373-L375

Reply:
> The decision was made to accept several types as input to the token list 

So we'd want

```python3
def add_token_list(self, tokens: Union[List[str], List[bytes], List[bytearray):
```

Correct? Or would you want to allow the items to be non-homogenous like `List[Union[str, bytes, bytearray]]`?

Reply:
What is the differences of the python types `str`, `bytes` and `bytearray`? If they all resolve to the same when written to disk then any of them could be used.

Reply:
Make it as simple it could be as long there is no difference in how the tokens are written to disk.

Reply:
str is unicode text which will be encoded as utf-8 before being written to disk. bytes and bytearray are binary data. Those two are subclasses of ByteString, but we can't use that because it also allows memoryview. The least repetitive way to write this would be to use a TypeVar:
```python
StrOrBytes = TypeVar('StrOrBytes', str, bytes, bytearray)
# ...
    def add_token_list(self, tokens: Sequence[StrOrBytes]):
        # ...
    def add_token_merges(self, merges: Sequence[StrOrBytes]):
        # ...
```

Reply:
Should merges actually support all of those?